File: src.net/Marketo-Rest/Api/BulkActivityExtract.cs
using Newtonsoft.Json.Linq;
using System;
using System.Threading.Tasks;

namespace Marketo.Api
{
    public class BulkActivityExtract
    {
        readonly Action<string> _log = util.logger;
        readonly MarketoClient _marketo;
        readonly Connection _connection;
        readonly Retry _retry;

        public BulkActivityExtract(MarketoClient marketo, Connection connection)
        {
            _marketo = marketo;
            _connection = connection;
            _retry = new Retry(new { maxRetries = 10, initialDelay = 30000, maxDelay = 60000 });
        }

        public Task<dynamic> Create(dynamic filter, dynamic options = null)
        {
            var path = util.createBulkPath("activities", "export", "create.json");
            options = dyn.exp(options);
            options.filter = filter;
            return _connection.postJson(path, options);
        }

        public Task<dynamic> Enqueue(string exportId, dynamic options = null)
        {
            var path = util.createPath("activities", "export", exportId, "enqueue.json");
            return _connection.post(path, new { data = options });
        }

        public Task<dynamic> Status(string exportId, dynamic options = null)
        {
            var path = util.createPath("activities", "export", exportId, "status.json");
            return _connection.get(path, new { data = options });
        }

        public async Task<dynamic> StatusTilCompleted(string exportId, dynamic options = null)
        {
            Func<bool, Task<JObject>> requestFn = async (forceOAuth) =>
            {
                dynamic data = await Status(exportId);
                if (!(bool)data.success)
                {
                    var msg = (string)data.errors[0].message;
                    _log(msg);
                    throw new MarketoException(msg);
                }
                Console.WriteLine($"STATUS: {data.result[0].status}");
                if (data.result[0].status == "Queued" || data.result[0].status == "Processing")
                    throw new MarketoException(null)
                    {
                        Id = data["requestId"],
                        Code = 606
                    };
                return data;
            };
            return await _retry.start(requestFn);
        }

        public Task<dynamic> Cancel(string exportId, dynamic options = null)
        {
            var path = util.createPath("activities", "export", exportId, "cancel.json");
            return _connection.post(path, new { data = options });
        }

        public async Task<dynamic> Get(dynamic filter, dynamic options = null)
        {
            var data = await Create(filter, options);
            if (!(bool)data.success)
            {
                var msg = (string)data.errors[0].message;
                _log(msg);
                throw new MarketoException(msg);
            }
            var exportId = (string)data.results[0].exportId;
            try
            {
                data = await Enqueue(exportId);
                if (!(bool)data.success)
                {
                    var msg = (string)data.errors[0].message;
                    _log(msg);
                    throw new MarketoException(msg);
                }
                data = await StatusTilCompleted(exportId);
            }
            catch (Exception err) { await Cancel(exportId); throw err; }
            return data;
        }

        public Task<dynamic> File(string exportId, dynamic options = null)
        {
            var path = util.createPath("activities", "export", exportId, "file.json");
            return _connection.get(path, new { data = options });
        }
    }
}

File: lib/api/bulkActivityExtract.js
var _ = require('lodash'),
  Promise = require('bluebird'),
  util = require('../util'),
  Retry = require('../retry'),
  BulkFileStream = require('../BulkFileStream'),
  log = util.logger();

function BulkActivityExtract(marketo, connection) {
  this._marketo = marketo;
  this._connection = connection;
  this._retry = new Retry({ maxRetries: 10, initialDelay: 30000, maxDelay: 60000 });
}

BulkActivityExtract.prototype = {
  create: function (filter, options) {
    var path = util.createBulkPath('activities', 'export', 'create.json');
    options = _.extend({}, options, {
      filter: filter,
    });
    return this._connection.postJson(path, options, { _method: 'POST' });
  },
  enqueue: function (exportId, options) {
    var path = util.createBulkPath('activities', 'export', exportId, 'enqueue.json');
    options = _.extend({}, options, {
      _method: 'POST'
    });
    return this._connection.post(path, { data: options });
  },
  status: function (exportId, options) {
    var path = util.createBulkPath('activities', 'export', exportId, 'status.json');
    options = _.extend({}, options, {
      _method: 'GET'
    });
    return this._connection.get(path, { data: options });
  },
  statusTillCompleted: function (exportId, options) {
    var self = this
    var requestFn = function () {
      return new Promise(function (resolve, reject) {
        self.status(exportId).then(function (data) {
          if (!data.success) {
            var msg = data.errors[0].message;
            log.error(msg);
            return reject(msg);
          }
          if (data.result[0].status == 'Queued' || data.result[0].status == 'Processing') {
            return reject({
              requestId: data.requestId,
              errors: [{ code: '606' }]
            });
          }
          resolve(data);
        }).catch(reject);
      })
    };
    return this._retry.start(requestFn, this);
  },
  cancel: function (exportId, options) {
    var path = util.createBulkPath('activities', 'export', exportId, 'cancel.json');
    options = _.extend({}, options, {
      _method: 'POST'
    });
    return this._connection.post(path, { data: options });
  },
  get: function (filter, options) {
    var self = this;
    return new Promise(function (resolve, reject) {
      self.create(filter, options).then(function (data) {
        if (!data.success) {
          var msg = data.errors[0].message;
          log.error(msg);
          return defer.reject(msg);
        }
        let exportId = data.result[0].exportId;
        self.enqueue(exportId).then(function (data) {
          if (!data.success) {
            var msg = data.errors[0].message;
            log.error(msg);
            return reject(msg);
          }
          self.statusTillCompleted(exportId)
            .then(resolve)
            .catch(reject);
        }).catch(function (err) {
          self.cancel(exportId).then(function () {
            return reject(err);
          }).catch(reject);
        });
      }).catch(reject);
    })
  },
  // FILE
  file: function (exportId, options) {
    var path = util.createBulkPath('activities', 'export', exportId, 'file.json');
    options = _.extend({}, options, {
      _method: 'GET'
    });
    return this._connection.get(path, { data: options });
  },
  fileStream: async function (exportId, fileSize, rangeSize) {
    var path = util.createBulkPath('activities', 'export', exportId, 'file.json');
    
    return new BulkFileStream(this._connection, path, fileSize, rangeSize);
  },
};

module.exports = BulkActivityExtract;
